File: app2/src/main/java/ru/neverlands/anclient/manager/KaznaItemDetailsCache.java
package ru.neverlands.anclient.manager;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.nio.charset.StandardCharsets;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import ru.neverlands.anclient.model.KaznaItemDetails;
import ru.neverlands.anclient.utils.AppLog;
import ru.neverlands.anclient.utils.AppVars;
import ru.neverlands.anclient.utils.GameInfoStorage;

/**
 * UTF-8 кеш UID-деталей казны в `info/<profile nick>/kazna/uids.txt`.
 *
 * Назначение после доработок:
 * - хранит свойства и image URL, которые сервер отдаёт в inventory-card
 *   `main.php?get_id=57&uid=...`, но не отдаёт в строках казны;
 * - пополняется из `InventoryParser.mainPhpInv(...)` и cache-only входов
 *   `InventoryParser.syncKaznaItemDetailsCacheFromHtml(...)`;
 * - читается `KaznaManager.loadItemDetailsByUid()` и передаётся в
 *   `KaznaItemAdapter.submitDetails(...)`;
 * - не является источником action UID для `Взять/Пожертвовать/Собрать`.
 *
 * Decision points:
 * - кеш пополняется только из уже загруженного HTML инвентаря;
 * - merge не затирает ранее известные поля пустыми значениями из неполной карточки;
 * - файл отделён от `info/<profile nick>/kazna/kazna.txt`, чтобы snapshot казны
 *   оставался только серверным списком;
 * - путь привязан к текущему профилю, чтобы детали вещей разных персонажей не
 *   смешивались между сессиями.
 */
public final class KaznaItemDetailsCache {
    private static final String TAG = "KaznaUidDetailsCache";
    private static final String TRACE = "KAZNA_TRACE";
    private static final String KAZNA_DIR = "kazna";
    private static final String UID_DETAILS_FILE = "uids.txt";
    private static final Object LOCK = new Object();

    private KaznaItemDetailsCache() {
    }

    public static File getCacheFile() {
        return getCacheFile(currentProfileNickForStorage());
    }

    public static File getCacheFile(String profileNick) {
        File dir = getKaznaDir(profileNick);
        return dir == null ? null : new File(dir, UID_DETAILS_FILE);
    }

    public static Map<String, KaznaItemDetails> loadAll() {
        return loadAll(currentProfileNickForStorage());
    }

    public static Map<String, KaznaItemDetails> loadAll(String profileNick) {
        synchronized (LOCK) {
            return loadAllLocked(profileNick);
        }
    }

    public static void mergeFromInventoryDetails(List<KaznaItemDetails> details) {
        mergeFromInventoryDetails(details, currentProfileNickForStorage());
    }

    public static void mergeFromInventoryDetails(List<KaznaItemDetails> details, String profileNick) {
        if (details == null || details.isEmpty()) {
            return;
        }
        synchronized (LOCK) {
            Map<String, KaznaItemDetails> current = loadAllLocked(profileNick);
            int parsed = 0;
            int changed = 0;
            for (KaznaItemDetails detail : details) {
                if (detail == null || detail.uid.isEmpty()) {
                    continue;
                }
                parsed++;
                KaznaItemDetails merged = merge(current.get(detail.uid), detail);
                if (!same(current.get(detail.uid), merged)) {
                    current.put(detail.uid, merged);
                    changed++;
                }
            }
            if (changed > 0) {
                saveAllLocked(current, profileNick);
                File file = getCacheFile(profileNick);
                AppLog.i(TAG, TRACE + " uid details updated: parsed=" + parsed
                        + ", changed=" + changed
                        + ", file=" + (file == null ? "" : file.getAbsolutePath()));
            }
        }
    }

    private static KaznaItemDetails merge(KaznaItemDetails oldValue, KaznaItemDetails newValue) {
        if (oldValue == null) {
            return new KaznaItemDetails(
                    newValue.uid,
                    newValue.name,
                    newValue.imageUrl,
                    newValue.propertiesText,
                    System.currentTimeMillis());
        }
        return new KaznaItemDetails(
                firstNonEmpty(newValue.uid, oldValue.uid),
                firstNonEmpty(newValue.name, oldValue.name),
                firstNonEmpty(newValue.imageUrl, oldValue.imageUrl),
                firstNonEmpty(newValue.propertiesText, oldValue.propertiesText),
                System.currentTimeMillis());
    }

    private static Map<String, KaznaItemDetails> loadAllLocked(String profileNick) {
        LinkedHashMap<String, KaznaItemDetails> result = new LinkedHashMap<>();
        File file = getCacheFile(profileNick);
        if (file == null || !file.exists()) {
            return result;
        }
        String json = readUtf8(file);
        if (json.isEmpty()) {
            return result;
        }
        try {
            JSONObject root = new JSONObject(json);
            JSONArray items = root.optJSONArray("items");
            if (items == null) {
                return result;
            }
            for (int index = 0; index < items.length(); index++) {
                JSONObject item = items.optJSONObject(index);
                if (item == null) {
                    continue;
                }
                KaznaItemDetails details = new KaznaItemDetails(
                        item.optString("uid", ""),
                        item.optString("name", ""),
                        item.optString("imageUrl", ""),
                        item.optString("propertiesText", ""),
                        item.optLong("updatedAtMs", 0L));
                if (!details.uid.isEmpty()) {
                    result.put(details.uid, details);
                }
            }
        } catch (Exception e) {
            AppLog.w(TAG, TRACE + " uid details read failed", e);
        }
        return result;
    }

    private static void saveAllLocked(Map<String, KaznaItemDetails> detailsByUid, String profileNick) {
        File file = getCacheFile(profileNick);
        if (file == null) {
            AppLog.w(TAG, TRACE + " uid details save skipped: info dir unavailable");
            return;
        }
        try {
            JSONObject root = new JSONObject();
            root.put("generatedAtMs", System.currentTimeMillis());
            JSONArray items = new JSONArray();
            for (KaznaItemDetails detail : detailsByUid.values()) {
                if (detail == null || detail.uid.isEmpty()) {
                    continue;
                }
                JSONObject item = new JSONObject();
                item.put("uid", detail.uid);
                item.put("name", detail.name);
                item.put("imageUrl", detail.imageUrl);
                item.put("propertiesText", detail.propertiesText);
                item.put("updatedAtMs", detail.updatedAtMs);
                items.put(item);
            }
            root.put("items", items);
            writeUtf8(file, root.toString(2));
        } catch (Exception e) {
            AppLog.e(TAG, TRACE + " uid details save failed", e);
        }
    }

    private static File getKaznaDir(String profileNick) {
        File dir = GameInfoStorage.resolveUserInfoSubDir(profileNick, KAZNA_DIR);
        if (dir == null && AppVars.getContext() != null) {
            File infoDir = new File(AppVars.getContext().getFilesDir(), "info");
            dir = new File(new File(infoDir, GameInfoStorage.sanitizeProfileName(profileNick)), KAZNA_DIR);
        }
        if (dir == null) {
            return null;
        }
        if (!dir.exists() && !dir.mkdirs()) {
            AppLog.w(TAG, TRACE + " failed to create uid details dir: " + dir.getAbsolutePath());
            return null;
        }
        return dir;
    }

    private static String currentProfileNickForStorage() {
        if (AppVars.Profile != null && AppVars.Profile.UserNick != null) {
            String nick = AppVars.Profile.UserNick.trim();
            if (!nick.isEmpty()) {
                return nick;
            }
        }
        String runtimeNick = AppVars.RuntimeAuthUserNick == null ? "" : AppVars.RuntimeAuthUserNick.trim();
        return runtimeNick.isEmpty() ? "profile" : runtimeNick;
    }

    private static String readUtf8(File file) {
        try (InputStreamReader reader = new InputStreamReader(new FileInputStream(file), StandardCharsets.UTF_8)) {
            StringBuilder sb = new StringBuilder();
            char[] buffer = new char[4096];
            int read;
            while ((read = reader.read(buffer)) != -1) {
                sb.append(buffer, 0, read);
            }
            return sb.toString();
        } catch (Exception e) {
            AppLog.w(TAG, TRACE + " uid details utf8 read failed: " + file.getAbsolutePath(), e);
            return "";
        }
    }

    private static void writeUtf8(File file, String content) throws Exception {
        File parent = file.getParentFile();
        if (parent != null && !parent.exists() && !parent.mkdirs()) {
            throw new IllegalStateException("Failed to create directory: " + parent.getAbsolutePath());
        }
        try (OutputStreamWriter writer = new OutputStreamWriter(new FileOutputStream(file), StandardCharsets.UTF_8)) {
            writer.write(content == null ? "" : content);
        }
    }

    private static String firstNonEmpty(String primary, String fallback) {
        return primary == null || primary.trim().isEmpty() ? safe(fallback) : primary.trim();
    }

    private static boolean same(KaznaItemDetails left, KaznaItemDetails right) {
        if (left == right) {
            return true;
        }
        if (left == null || right == null) {
            return false;
        }
        return safe(left.uid).equals(safe(right.uid))
                && safe(left.name).equals(safe(right.name))
                && safe(left.imageUrl).equals(safe(right.imageUrl))
                && safe(left.propertiesText).equals(safe(right.propertiesText));
    }

    private static String safe(String value) {
        return value == null ? "" : value;
    }
}

File: ANClient/Kazna/KaznaItemDetailsCache.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Windows.Forms;
using Newtonsoft.Json.Linq;

namespace ANClient.Kazna
{
    internal static class KaznaItemDetailsCache
    {
        private const string TraceChain = "KAZNA_TRACE";
        private static readonly object LockObject = new object();

        internal static Dictionary<string, KaznaItemDetails> LoadAll()
        {
            lock (LockObject)
            {
                return LoadAllLocked(CurrentProfileNickForStorage());
            }
        }

        internal static void MergeFromInventoryDetails(List<KaznaItemDetails> details)
        {
            if (details == null || details.Count == 0)
                return;

            lock (LockObject)
            {
                var profileNick = CurrentProfileNickForStorage();
                var current = LoadAllLocked(profileNick);
                var parsed = 0;
                var changed = 0;
                for (var i = 0; i < details.Count; i++)
                {
                    var detail = details[i];
                    if (detail == null || string.IsNullOrEmpty(detail.Uid))
                        continue;

                    parsed++;
                    KaznaItemDetails oldValue;
                    current.TryGetValue(detail.Uid, out oldValue);
                    var merged = Merge(oldValue, detail);
                    if (!Same(oldValue, merged))
                    {
                        current[merged.Uid] = merged;
                        changed++;
                    }
                }

                if (changed > 0)
                {
                    SaveAllLocked(current, profileNick);
                    AppLog.i(TraceChain, "KaznaItemDetailsCache", "uid details updated: parsed=" + parsed + ", changed=" + changed + ", file=" + GetCacheFile(profileNick));
                }
            }
        }

        private static KaznaItemDetails Merge(KaznaItemDetails oldValue, KaznaItemDetails newValue)
        {
            if (oldValue == null)
                return new KaznaItemDetails(newValue.Uid, newValue.Name, newValue.ImageUrl, newValue.PropertiesText, DateTime.Now.Ticks);

            return new KaznaItemDetails(
                FirstNonEmpty(newValue.Uid, oldValue.Uid),
                FirstNonEmpty(newValue.Name, oldValue.Name),
                FirstNonEmpty(newValue.ImageUrl, oldValue.ImageUrl),
                FirstNonEmpty(newValue.PropertiesText, oldValue.PropertiesText),
                DateTime.Now.Ticks);
        }

        private static Dictionary<string, KaznaItemDetails> LoadAllLocked(string profileNick)
        {
            var result = new Dictionary<string, KaznaItemDetails>(StringComparer.OrdinalIgnoreCase);
            var file = GetCacheFile(profileNick);
            if (!File.Exists(file))
                return result;

            try
            {
                var root = JObject.Parse(File.ReadAllText(file, System.Text.Encoding.UTF8));
                var items = root["items"] as JArray;
                if (items == null)
                    return result;

                for (var i = 0; i < items.Count; i++)
                {
                    var item = items[i] as JObject;
                    if (item == null)
                        continue;

                    var detail = new KaznaItemDetails(
                        (string)item["uid"],
                        (string)item["name"],
                        (string)item["imageUrl"],
                        (string)item["propertiesText"],
                        (long?)item["updatedAtTicks"] ?? 0L);
                    if (!string.IsNullOrEmpty(detail.Uid))
                        result[detail.Uid] = detail;
                }
            }
            catch (Exception ex)
            {
                AppLog.w(TraceChain, "KaznaItemDetailsCache", "uid details read failed", ex);
            }

            return result;
        }

        private static void SaveAllLocked(Dictionary<string, KaznaItemDetails> detailsByUid, string profileNick)
        {
            try
            {
                var file = GetCacheFile(profileNick);
                var dir = Path.GetDirectoryName(file);
                if (!Directory.Exists(dir))
                    Directory.CreateDirectory(dir);

                var root = new JObject();
                root["generatedAtTicks"] = DateTime.Now.Ticks;
                var items = new JArray();
                foreach (var detail in detailsByUid.Values)
                {
                    if (detail == null || string.IsNullOrEmpty(detail.Uid))
                        continue;

                    var item = new JObject();
                    item["uid"] = detail.Uid;
                    item["name"] = detail.Name;
                    item["imageUrl"] = detail.ImageUrl;
                    item["propertiesText"] = detail.PropertiesText;
                    item["updatedAtTicks"] = detail.UpdatedAtTicks;
                    items.Add(item);
                }

                root["items"] = items;
                File.WriteAllText(file, root.ToString(), System.Text.Encoding.UTF8);
            }
            catch (Exception ex)
            {
                AppLog.e(TraceChain, "KaznaItemDetailsCache", "uid details save failed", ex);
            }
        }

        private static string GetCacheFile(string profileNick)
        {
            return Path.Combine(Path.Combine(Path.Combine(Application.StartupPath, "info"), SanitizeProfileName(profileNick)), "kazna" + Path.DirectorySeparatorChar + "uids.txt");
        }

        private static string CurrentProfileNickForStorage()
        {
            if (AppVars.Profile != null && !string.IsNullOrEmpty(AppVars.Profile.UserNick))
                return AppVars.Profile.UserNick.Trim();

            return "profile";
        }

        private static string SanitizeProfileName(string value)
        {
            var safe = string.IsNullOrEmpty(value) ? "profile" : value.Trim();
            var invalid = Path.GetInvalidFileNameChars();
            for (var i = 0; i < invalid.Length; i++)
                safe = safe.Replace(invalid[i], '_');

            return string.IsNullOrEmpty(safe) ? "profile" : safe;
        }

        private static string FirstNonEmpty(string first, string second)
        {
            return string.IsNullOrEmpty(first) ? (second ?? string.Empty) : first;
        }

        private static bool Same(KaznaItemDetails first, KaznaItemDetails second)
        {
            if (first == null || second == null)
                return first == second;

            return string.Equals(first.Uid, second.Uid, StringComparison.Ordinal) &&
                   string.Equals(first.Name, second.Name, StringComparison.Ordinal) &&
                   string.Equals(first.ImageUrl, second.ImageUrl, StringComparison.Ordinal) &&
                   string.Equals(first.PropertiesText, second.PropertiesText, StringComparison.Ordinal);
        }
    }
}
